### ISSUE TITLE
bnd-maven-plugin: configuration parameters bnd and bndfile are not evaluated in execution level configuration

### ISSUE THREAD

**Issue by kwin** [bndtools/bnd#1771]
Due to the way how bnd-maven-plugin evaluates both configuration parameters `bnd` and `bndfile` only plugin level configuration is supported but not execution level configuration (https://maven.apache.org/guides/mini/guide-default-execution-ids.html). For example the following configuration is not considered:

```
<plugin>
  <groupId>biz.aQute.bnd</groupId>
  <artifactId>bnd-maven-plugin</artifactId>
  <version>3.3.0</version>
  <executions>
    <execution>
      <goals>
        <goal>bnd-process</goal>
      </goals>
       <configuration>
	   <bnd><![CDATA[
			testheader:testvalue]]></bnd>
    </execution>
  </executions>
<plugin>
```

Also code completion (like provided in m2e since 1.6, http://www.eclipse.org/m2e/documentation/release-notes-16.html#improved-auto-45-completion-in-pom-46-xml-editor) is not supported, because the parameters do not appear in the `META-INF\maven\plugin.xml`.

Therefore please rather use the default way of injecting those configuration parameters into the plugin, rather than relying on evaluating the configuration from the pom.xml on your own.

**Comment by bjhargrave** [comment]
I don't fully understand the difference as I am not a maven expert. The bnd-maven-plugin supports merging the configuration information via parent poms which is why the code is the way it is. If you can make a PR to also address your execution thing, please do.

Also for m2e code completion, that sounds like a different and unrelated issue. A separate issue would be better.

**Comment by kwin** [comment]
Actually both issues are caused by the fact how you evaluate the configuration. I will try to come up with a PR though.

**Comment by bjhargrave** [comment]
Great. Please make sure to add additional tests to validate the PR changes (so we don't break it in the future.)

**Comment by bjhargrave** [comment]
The plugin recurses to the deepest ancestor pom and collects its configuration. As the recursion unwinds, the configuration of each pom is overlaid. Using normal configuration injection will only handle the case of the current pom's configuration and not properly overlay the configurations of the hierarchy of poms. Study the `loadProjectProperties` method carefully.

**Comment by kwin** [comment]
The inheritance model is not 100% Maven compliant (https://maven.apache.org/pom.html#Plugins). E.g. when using in-pom bnd instructions through parameter `bnd` it considers basically the bnd-instructions from the parent and merges those with the ones from the child on a key-value level (whereby the child may overwrite properties of the same name from the parent). Also right now a `bndfile` in a child would not completely overwrite `bnd` instructions from the parent, but would rather be merged with those.
The Maven approach would be: In case the same plugin configuration element (e.g. `bnd`) is defined on both parent and child, the one from the child overwrites (completely) the one from the parent. Also for the `bndfile` parameter I would assume, that a child can completely overwrite the `bndfile` (even the default one) from the parent, so that this would then no longer be active but rather only the `bndfile` being referenced in the current pom.

This rather simplistic inheritance would just allow inheritance on a `all or nothing` level, but it would comply with Maven standards.

If that really is deliberate behaviour it should be documented better in the readme and also in the description of the configuration parameters (i.e. javadoc on the `@Parameter` annotation), otherwise I would rather stick to the simple Maven inheritance model.

**Comment by rotty3000** [comment]
But merge properties is a native feature of bnd, it has little to do with
maven, the support for properties in parent projects in maven for bnd is a
bnd feature and it's an important one for me and for many other bnd plugin
users. It's one we've been using within our other bnd usages forever. At
best I could see having an opt-out model but I doubt that you'll convince
anyone as a whole to change the merge model to be all of nothing.

**Comment by kwin** [comment]
@bjhargrave Could you please reopen this issue as the merged PR was reverted in #1781?
